zlib: do not coalesce multiple `.flush()` calls

addaleax opened the pull request:
This is an approach to address the issue linked below. Previously,
when `.write()` and `.flush()` calls to a zlib stream were interleaved
synchronously (i.e. without waiting for these operations to finish),
multiple flush calls would have been coalesced into a single flushing
operation.

This patch changes behaviour so that each `.flush()` all corresponds
to one flushing operation on the underlying zlib resource, and the
order of operations is as if the `.flush()` call were a `.write()`
call.

One test had to be removed because it specifically tested the previous
behaviour.

As a drive-by fix, this also makes sure that all flush callbacks are
called. Previously, that was not the case.

/cc @mscdex – is this something you prefer over the current behaviour?

Fixes: https://github.com/nodejs/node/issues/28478

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

lpinca commented on test/parallel/test-zlib-flush-write-sync-interleaved.js:
```diff
@@ -0,0 +1,39 @@
+    { written: 'abc' },
+    'flushed',
+    { written: 'def' },
+    { read: 'abc' },
```
I wonder if it's better to use a dedicated array for `decompress.write()` in order to not have a flaky test, or is this order guaranteed to always be respected?

addaleax commented:
I’ve updated the zlib source to use a fixed number of `Buffer` objects for flushing operations instead of allocating new ones each time, and refactored the test file following @lpinca’s good observation that the test could have been flaky in its previous form.

@lpinca @trott @BridgeAR I’d appreciate it if you could take another look!

addaleax commented:
Landed in 518ffc125680f0916635d2ed97c076dbff3bd05b
